### PR TITLE
Chapter0

### PR DESCRIPTION
**Pull request by manojdvs7**

### CODE REVIEW

File: C-Chapter0/src/count.cpp [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+int count(int num1, int num2) {
+	if (num1 == 0)
+		return -1;
+	if (num2 == 0)
+		return 0;
+	if (num1 >= num2)
+		return 1;
+	if (num2%num1 == 0)
+		return 1 + count(num1, num2 - 1);
+
```
Comment (krushiraj) [inline C-Chapter0/src/count.cpp:23]: Good to see that you're using recursion for this. Nice approach. Can you discuss any other approaches that you've thought of here in comments.

File: C-Chapter0/src/countFactorials.cpp [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+}
+
+int countFactorials(long l, long h)
+{
+	if (l > h || l<0 || h<0)
+		return -1;
+	int count = 0,i=1;
+	while (1)
+	{
+		if (fact(i) >= l && fact(i) <= h)
```
Comment (krushiraj) [inline C-Chapter0/src/countFactorials.cpp:36]: This seems to be a very naive approach. Can you think of any optimizations? I want you to come up with the drawbacks of this approach first then discuss the solution for each of the drawbacks.

File: C-Chapter0/src/countOnes.cpp [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+	long long n = 111;
+	int count = 3;
+	while (n % number != 0)
+	{
+
+		n = n * 10 + 1;
+			printf("\n%d", n);
+		count++;
+	}
+
```
Comment (krushiraj) [inline C-Chapter0/src/countOnes.cpp:40]: Good one. Think of how can we solve this for numbers that can't fit into `long long`.

File: C-Chapter0/src/isOlder.cpp [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+		return true;
+}
+
+int isOlder(char *dob1, char *dob2) {
+	bool flag = checkValid(dob1, dob2);
+	if (flag)
+		return stringComp(dob1,dob2);
+	else
+		return -1;
+}
```
Comment (krushiraj) [inline C-Chapter0/src/isOlder.cpp:69]: Modularity 👌

File: C-Chapter0/src/nthNumber.cpp [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+	}
+	return ul;
+}
+int isprime(int n)
+{
+	int count = 0;
+	for (int i = 1; i <= n; i++)
+	{
+		if (n%i == 0)
+			count++;
```
Comment (krushiraj) [inline C-Chapter0/src/nthNumber.cpp:44]: I guess you can optimize this `isprime` to some better extent. Think of it and discuss where you can optimize~

File: C-Chapter0/src/nthNumber.cpp [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+	else
+		return 0;
+}
+int nthPrime(int num)
+{
+	if (num <= 0)
+		return -1;
+	int prime;
+	int num1 = 2;
+	while (num)
```
Comment (krushiraj) [inline C-Chapter0/src/nthNumber.cpp:60]: Here also, you can add some optimizations to cut of some CPU utilization.

File: C-Chapter0/src/sum.cpp [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+
+ERROR CASES:
+
+NOTES:
+*/
+
+int sum(int a, int b)
+{
+	return a + b;
+	return 0;
```
Comment (krushiraj) [inline C-Chapter0/src/sum.cpp]: Why there are 2 return statements?

**Review by krushiraj — CHANGES_REQUESTED** [review]

File: C-Chapter0/src/count.cpp (hunk 1 above)
Comment (manojdvs7) [reply C-Chapter0/src/count.cpp:23]: An another approach I thought was of course a loop solution.

Comment (krushiraj) [reply C-Chapter0/src/count.cpp:23]: Don't go by logic. Think if you can have any mathematical solution that can get you the answer by O(1)?

File: C-Chapter0/src/countFactorials.cpp (hunk 2 above)
Comment (manojdvs7) [reply C-Chapter0/src/countFactorials.cpp:36]: Drawbacks:
The fact( ) is called starting with i=1 till it exceeds the range,even when fact(1) is not in the range(first factorial in the range).

File: C-Chapter0/src/countOnes.cpp (hunk 3 above)
Comment (manojdvs7) [reply C-Chapter0/src/countOnes.cpp:40]: Yes.I was thinking, that number can be divided into small parts and can be combined at the end.
At last I am still facing overflow  condition while combining.

File: C-Chapter0/src/isOlder.cpp (hunk 4 above)
Comment (manojdvs7) [reply C-Chapter0/src/isOlder.cpp:69]: Thanks Bro :)

File: C-Chapter0/src/nthNumber.cpp (hunk 5 above)
Comment (manojdvs7) [reply C-Chapter0/src/nthNumber.cpp:44]: Yes. condition(i<=n/2).I forgot.

File: C-Chapter0/src/sum.cpp (hunk 7 above)
Comment (manojdvs7) [reply C-Chapter0/src/sum.cpp]: I am sorry, I din't see that.

File: C-Chapter0/src/nthNumber.cpp (hunk 6 above)
Comment (manojdvs7) [reply C-Chapter0/src/nthNumber.cpp:60]: By removing prime only bro?

Comment (krushiraj) [reply C-Chapter0/src/nthNumber.cpp:60]: No even number is prime except for 2. Can you make use of this fact and optimize?

**Review by krushiraj — APPROVED** [review]

**Comment by krushiraj** [comment]
I'm not enforcing you to make few optimisation to higher extent because this is getting started lesson. But make sure that from next lesson onwards write the most optimised version of an algo and raise a PR.

**Comment by manojdvs7** [comment]
> I'm not enforcing you to make few optimisation to higher extent because this is getting started lesson. But make sure that from next lesson onwards write the most optimised version of an algo and raise a PR.

Yes bro.I will try my level best.

**Pull request merged**